Create Migration Guide (#103)
LiskHQ/lisk-migrator: docs/migration.md

> # Lisk V4.x Migration guide

Reviewer: ```suggestion
# Migration Guide
```

> This section explains how to migrate a Lisk Core v3.x node to Lisk Core v4.x using the Lisk migrator.

Reviewer: ```suggestion
This section explains how to migrate a Lisk Core v3.0.5 (or later) node to Lisk Core v4.x using the Lisk Migrator.
```

> The Lisk migrator CLI tool will generate a new genesis block for Lisk Core v4.

Reviewer: ```suggestion
The Lisk Migrator CLI tool will generate a new genesis block for Lisk Core v4.
```

> The new genesis block is created based on a snapshot of the old blockchain (running on Lisk Core v3.x) at a given height.

Reviewer: ```suggestion
The new genesis block is created based on a snapshot of the existing blockchain (running on Lisk Core v3.0.5+) at a given height.
```

> Note: All actively forging validators on the Lisk Mainnet and the Lisk Testnet need to follow this guide to correctly migrate their nodes to the new network, in order to not miss any blocks after the network hard fork.

Please ensure you are running version 3.x of Lisk Core to be able to seamlessly migrate to Lisk Core 4.x.

> ## Setting up the Lisk migrator

Reviewer: ```suggestion
## Setting up the Lisk Migrator
```

### Download checksum and verify

Download the checksum and verify the successful download of lisk-migrator.

```

> curl -o lisk-migrator-v2.0.0.SHA256 https://downloads.lisk.com/lisk-migrator/lisk-migrator-v2.0.0.SHA256

Reviewer: ```suggestion
curl -O https://downloads.lisk.com/lisk-migrator/lisk-migrator-v2.0.0.SHA256
```

```

### Extract and add to PATH

```

> tar -xf ./lisk-migrator-v2.0.0.SHA256

Reviewer: Seems incorrect. There's no need to extract the checksum file. It must be used to validate the binary checksum.

On Linux:
```
sha256sum -c genesis_block.json.tar.gz.SHA256 genesis_block.json.tar.gz
genesis_block.json.tar.gz: OK
sha256sum: genesis_block.json.tar.gz: no properly formatted SHA256 checksum lines found
```

On Mac:
```
shasum -a 256 -c genesis_block.json.tar.gz.SHA256 genesis_block.json.tar.gz
genesis_block.json.tar.gz: OK
shasum: genesis_block.json.tar.gz: no properly formatted SHA checksum lines found
```

```

Make the `lisk-migrator` command available in the PATH, e.g. by executing the following command:

```
export PATH="$PATH:$HOME/lisk-migrator/bin"
```

> Replace `$HOME` with the absolute path of where the `lisk-migrator` folder is located, in case it was extracted somewhere else other than in your home directory.

> Alternatively migrator setup can be done by following the steps defined in section [here](../README.md).

## Migration Steps

**Check the announced snapshot height**

- For Mainnet: `TBD`
- For Testnet: `TBD`

### Run Lisk Migrator

> The Lisk migrator v2.0.0 also allow users to download and start Lisk Core v4.x automatically after the migration. It can be done by passing the relavant flags.

Reviewer: ```suggestion
The Lisk Migrator v2 also allow users to download and start the Lisk Core v4.x automatically post migration. It can be done by passing the relevant flags.
```

  --auto-migrate-config                                Migrate user configuration automatically. Default to false.
  --auto-start-lisk-core-v4                            Start lisk core v4 automatically. Default to false.
  --snapshot-time-gap=snapshot-time-gap                The number of seconds elapsed between the block at height HEIGHT_SNAPSHOT and the snapshot block.
  --use-existing-snapshot                              Use existing database snapshot (Temporary flag, will be removed once createSnapshot command is available on Lisk Core v3.x).

EXAMPLES
  lisk-migrator --snapshot-path  /path/to/snapshot  --snapshot-height 20931763 --lisk-core-path /path/to/data-dir

  lisk-migrator --snapshot-path  /path/to/snapshot  --snapshot-height 20931763 --lisk-core-path /path/to/data-dir --auto-download-lisk-core-v4 --auto-start-lisk-core-v4 --auto-migrate-config
```

If you have added `lisk-migrator` to the PATH as described in the section [setting-up-the-lisk-migrator](#setting-up-the-lisk-migrator), you can start the migration script by running the following command in the terminal:

**Mainnet**

```

> lisk-migrator --snapshot-height ${snapshotHeight} --output ~/.lisk/lisk-core/config/mainnet --lisk-core-v3-data-path ~/lisk-main --auto-download-lisk-core-v4  --auto-migrate-config --auto-start-lisk-core-v4

Reviewer: ```suggestion
lisk-migrator --snapshot-height [recommendedSnapshotHeight] --output ~/.lisk/lisk-core/config/mainnet --lisk-core-v3-data-path ~/lisk-main --auto-download-lisk-core-v4  --auto-migrate-config --auto-start-lisk-core-v4
```

```

**Testnet**

```

> lisk-migrator --snapshot-height ${snapshotHeight} --output ~/.lisk/lisk-core/config/testnet --lisk-core-v3-data-path ~/lisk-test --auto-download-lisk-core-v4  --auto-migrate-config --auto-start-lisk-core-v4

Reviewer: ```suggestion
lisk-migrator --snapshot-height [recommendedSnapshotHeight] --output ~/.lisk/lisk-core/config/testnet --lisk-core-v3-data-path ~/lisk-test --auto-download-lisk-core-v4  --auto-migrate-config --auto-start-lisk-core-v4
```

```

- `--snapshot-height`:
  The height on which the blockchain snapshot will be performed.
  The snapshot height will be announced separately.
- `--output`:
  The absolute path to the directory, where the newly generated genesis block should be saved.
- `--lisk-core-v3-data-path`:
  The absolute path to the directory, where the Lisk Core v3.x node is located.
- `--auto-download-lisk-core-v4`:
  Download Lisk Core v4.x automatically.
- `--auto-migrate-config`:
  Migrate Lisk Core v3.x configuration to v4.x automatically.
- `--auto-start-lisk-core-v4`:
  Start Lisk Core v4.x automatically.

Alternatively genesis block and configuration for Lisk Core v4.x migration can be created separately without starting Lisk Core v4.x automatically:

**Mainnet**

```

> lisk-migrator --snapshot-height ${snapshotHeight} --output ~/.lisk/lisk-core/config/mainnet --lisk-core-v3-data-path ~/lisk-main --auto-migrate-config
> ```
> 
> **Testnet**
> 
> ```
> lisk-migrator --snapshot-height ${snapshotHeight} --output ~/.lisk/lisk-core/config/testnet --lisk-core-v3-data-path ~/lisk-test --auto-migrate-config

Reviewer: ```suggestion
lisk-migrator --snapshot-height [recommendedSnapshotHeight] --output ~/.lisk/lisk-core/config/mainnet --lisk-core-v3-data-path ~/lisk-main --auto-migrate-config
```

**Testnet**

```
lisk-migrator --snapshot-height [recommendedSnapshotHeight] --output ~/.lisk/lisk-core/config/testnet --lisk-core-v3-data-path ~/lisk-test --auto-migrate-config
```

  -v, --version                                        show CLI version
  --auto-download-lisk-core-v4                         Download lisk core v4 automatically. Default to false.
  --auto-migrate-config                                Migrate user configuration automatically. Default to false.
  --auto-start-lisk-core-v4                            Start lisk core v4 automatically. Default to false.
  --snapshot-time-gap=snapshot-time-gap                The number of seconds elapsed between the block at height HEIGHT_SNAPSHOT and the snapshot block.
  --use-existing-snapshot                              Use existing database snapshot (Temporary flag, will be removed once createSnapshot command is available on Lisk Core v3.x).

EXAMPLES
  lisk-migrator --snapshot-path  /path/to/snapshot  --snapshot-height 20931763 --lisk-core-path /path/to/data-dir

  lisk-migrator --snapshot-path  /path/to/snapshot  --snapshot-height 20931763 --lisk-core-path /path/to/data-dir --auto-download-lisk-core-v4 --auto-start-lisk-core-v4 --auto-migrate-config
```

> If you have added `lisk-migrator` to the PATH as described in the section [setting-up-the-lisk-migrator](#setting-up-the-lisk-migrator), you can start the migration script by running the following command in the terminal:

Reviewer 2: ```suggestion
If you have added `lisk-migrator` to the PATH as described in the [setting-up-the-lisk-migrator](#setting-up-the-lisk-migrator) section, you can start the migration script by running the following command in the terminal:
```

**Mainnet**

```
lisk-migrator --snapshot-height [recommendedSnapshotHeight] --output ~/.lisk/lisk-core/config/mainnet --lisk-core-v3-data-path ~/lisk-main --auto-download-lisk-core-v4  --auto-migrate-config --auto-start-lisk-core-v4
```

**Testnet**

```
lisk-migrator --snapshot-height [recommendedSnapshotHeight] --output ~/.lisk/lisk-core/config/testnet --lisk-core-v3-data-path ~/lisk-test --auto-download-lisk-core-v4  --auto-migrate-config --auto-start-lisk-core-v4
```

- `--snapshot-height`:
  The height on which the blockchain snapshot will be performed.
  The snapshot height will be announced separately.
- `--output`:
  The absolute path to the directory, where the newly generated genesis block should be saved.
- `--lisk-core-v3-data-path`:
  The absolute path to the directory, where the Lisk Core v3.x node is located.
- `--auto-download-lisk-core-v4`:
  Download Lisk Core v4.x automatically.
- `--auto-migrate-config`:
  Migrate Lisk Core v3.x configuration to v4.x automatically.
- `--auto-start-lisk-core-v4`:
  Start Lisk Core v4.x automatically.

> Alternatively genesis block and configuration for Lisk Core v4.x migration can be created separately without starting Lisk Core v4.x automatically:

Reviewer 2: ```suggestion
Alternatively, the genesis block and configuration for Lisk Core v4.x migration can be created separately without starting Lisk Core v4.x automatically:
```

**Mainnet**

```
lisk-migrator --snapshot-height [recommendedSnapshotHeight] --output ~/.lisk/lisk-core/config/mainnet --lisk-core-v3-data-path ~/lisk-main --auto-migrate-config
```

**Testnet**

```
lisk-migrator --snapshot-height [recommendedSnapshotHeight] --output ~/.lisk/lisk-core/config/testnet --lisk-core-v3-data-path ~/lisk-test --auto-migrate-config
```

> In case `--auto-start-lisk-core-v4` is disable, please install & start Lisk Core v4.x manually.

Reviewer 2: ```suggestion
In case `--auto-start-lisk-core-v4` is disabled, please install & start Lisk Core v4.x manually.
```

> Please follow [these](https://github.com/LiskHQ/lisk-core/blob/development/README.md#installation) steps for installation.

Reviewer 2: ```suggestion
Please follow the steps in the [README guide](https://github.com/LiskHQ/lisk-core/blob/development/README.md#installation) to perform the installation.
```
